homeassistant/components/roborock/image.py

```diff
@@ -130,29 +158,53 @@ async def create_coordinator_maps(
     )
-    for map_flag, map_info in maps_info:
-        # Load the map - so we can access it with get_map_v1
-        if map_flag != cur_map:
-            # Only change the map and sleep if we have multiple maps.
-            await coord.api.send_command(RoborockCommand.LOAD_MULTI_MAP, [map_flag])
-            coord.current_map = map_flag
-            # We cannot get the map until the roborock servers fully process the
-            # map change.
-            await asyncio.sleep(MAP_SLEEP)
-        # Get the map data
-        map_update = await asyncio.gather(
-            *[coord.cloud_api.get_map_v1(), coord.get_rooms()]
+    maps = await hass.async_add_executor_job(
+        roborock_storage.async_load_maps,
+        [roborock_map.name for roborock_map in coord.maps.values()],
+    )
+    storage_updates = []
+    for (map_flag, map_info), storage_map in zip(maps_info, maps):
+        unique_id = (
+            f"{slugify(coord.roborock_device_info.device.duid)}_map_{map_info.name}"
         )
-        api_data: bytes = map_update[0]
-        entities.append(
-            RoborockMap(
-                f"{slugify(coord.roborock_device_info.device.duid)}_map_{map_info.name}",
-                coord,
-                map_flag,
-                api_data,
-                map_info.name,
+        # Load the map - so we can access it with get_map_v1
+        api_data: bytes | None = storage_map
+        create_map = False
+        if api_data is None:
+            # Only get the map data on startup if a) we haven't added the entity before
+            # b) The entity does not have the needed restore data.
+            if map_flag != cur_map:
+                # Only change the map and sleep if we have multiple maps.
+                await coord.api.send_command(RoborockCommand.LOAD_MULTI_MAP, [map_flag])
+                coord.current_map = map_flag
+                # We cannot get the map until the roborock servers fully process the
+                # map change.
+                await asyncio.sleep(MAP_SLEEP)
+            # Get the map data
+            map_update = await asyncio.gather(
+                *[coord.cloud_api.get_map_v1(), coord.get_rooms()]
             )
+            api_data = map_update[0]
+            if api_data is None:
+                # If we fail to get the map data, we should set it to empty bytes,
+                # so it is setup, but set as unavailable
+                api_data = b""
+            create_map = True
+        roborock_map = RoborockMap(
+            unique_id,
+            coord,
+            map_flag,
+            api_data,
+            map_info.name,
+            roborock_storage,
+            create_map,
         )
+        entities.append(roborock_map)
+        if create_map and roborock_map.cached_map != b"":
+            storage_updates.append(
+                roborock_storage.async_save_map(map_info.name, roborock_map.cached_map)
+            )
+    await asyncio.gather(*storage_updates)
```

Comment:
It's better to make a function `async_save_maps` to avoid multiple tasks each scheduling their own executor job. Instead, have it all be done in 1 executor job.

Reply:
Also, any reason this needs to be happening in the "hot path" during startup? Can we do the saving in a background task (`config_entry.async_create_background_task`)

Reply:
Okay I believe I did it right - but you should probably double check my async_save_maps logic to make sure it's okay as a background task.